pyfive/dataobjects.py

```diff
@@ -659,7 +595,202 @@ def _decode_link_info_msg(data, offset):
+                return np.array([Reference(addr) for addr in ref_addresses])[args]
+            else:
+                raise NotImplementedError('datatype not implemented')
+
+
+
+    def iterchunks(self, sel=None):
+        """ 
+        Iterate over chunks in a chunked dataset. 
+        The optional sel argument is a slice or tuple of slices that defines the region to be used. #FIXME: sel not yet implemented.
+        If not set, the entire dataspace will be used for the iterator.
+        For each chunk within the given region, the iterator yields a tuple of slices that gives the intersection of the given chunk 
+        with the selection area. This can be used to read or write data in that chunk.
+        A TypeError will be raised if the dataset is not chunked.
+        A ValueError will be raised if the selection region is invalid.
+        """
+        if self.chunks is None:
+            raise TypeError('Dataset is not chunked')
+        return self.id._iter_chunks(sel)
+
+    def _get_selection_from_contiguous(self, args=None):
+        """
+        Two options, we either read the entire contiguous array, and pull out
+        the selection (args) from that, or we try and read the contiguous
+        array in pseudo chunks (being the contiguous data which varies on
+        the slowest axis). We only do the latter if a) we have the
+        pseudo_chunking turned on (see init method), b) the array is 
+        multi dimensional, and c) the slabs are big enough to bother
+        (i.e bigger than pseudo_block_size_kib). We use the fact that the
+        storage order is known (HDF writes in the C-order, last axis
+        varying fastest).
+        """
+
+        # don't want to be doing this if we are actually chunked!
+        assert self.chunks is None
+        
+        data_offset, = struct.unpack_from('<Q', self.msg_data, self.property_offset)
+        itemsize = np.dtype(self.dtype).itemsize
+        
+        #are we 1d, too small to worry about pseudo chunks, or is pseudo chunking turned off?
+        #if so, read the lot as one chunk, otherwise use collapse all but last dimension into chunks
+      
+        if len(self.shape) == 1:
+            self.pseudo_chunking = False
+
+        if self.pseudo_chunking:
+            stride = np.prod(self.shape[1:])*itemsize
+            if stride < self.pseudo_block_size_kib*1024:
+                self.pseudo_chunking = False
+        
+        if self.pseudo_chunking:
+            pseudo_chunks = np.copy(self.shape)
+            pseudo_chunks[0] = 1
+        else:
+            stride = np.prod(self.shape)*itemsize
+            pseudo_chunks=self.shape
+ 
+        if args is None:
+
+            # we need it all, let's get it all
+            self.fh.seek(data_offset)
+            chunk_buffer = self.fh.read(stride)
+            chunk_data = np.frombuffer(chunk_buffer, dtype=self.dtype)
+            chunk_data.reshape(self.shape, order=self.order)
```

Comment:
nice but reshaping tends to be fragile, can we do a wee try and except perhaps?

Reply:
If this is going to barf, I'd like it to barf up the entire stack, as there is something fundamentally wrong if the file has a shape that doesn't work ... _after_ it has been written.